server/backup_runner.go

```diff
@@ -50,15 +53,78 @@ func (s *BackupRunner) Run(ctx context.Context, wg *sync.WaitGroup) {
 						log.Error("Auto backup runner PANIC RECOVER", zap.Error(err))
 					}
 				}()
-				s.startAutoBackups(ctx, runningTasks, &mu)
+				s.startAutoBackups(ctx, runningBackupTasks, &muBackup)
+				s.downloadBinlogFiles(ctx)
 			}()
 		case <-ctx.Done(): // if cancel() execute
 			return
 		}
 	}
 }
 
-func (s *BackupRunner) startAutoBackups(ctx context.Context, runningTasks map[int]bool, mu *sync.RWMutex) {
+func (s *BackupRunner) downloadBinlogFiles(ctx context.Context) {
+	normal := api.Normal
+	instanceFind := &api.InstanceFind{RowStatus: &normal}
+	instanceList, err := s.server.store.FindInstance(ctx, instanceFind)
+	if err != nil {
+		log.Error("Failed to retrieve instance list", zap.Error(err))
+		return
+	}
+	downloadTimeout := s.backupRunnerInterval / 2
+	if downloadTimeout < time.Minute {
+		downloadTimeout = time.Minute
+	}
+	for _, instance := range instanceList {
+		if instance.Engine == db.MySQL && instance.RowStatus == api.Normal {
+			log.Debug("Checking if any databases of instance enables backup plan policy", zap.String("instance", instance.Name))
+			enabled, err := s.server.store.SomeDatabaseEnablesBackupPolicyInInstance(ctx, instance.ID)
+			if err != nil {
+				log.Error("Failed to get backup plan policy for databases of instance", zap.Error(err))
+				continue
+			}
+			if !enabled {
+				log.Debug("No database in instance enables backup plan policy, skip instance", zap.String("instance", instance.Name))
+				continue
+			}
+			// We do not cancel ctxDownload explicitly, because the downloading goroutine may run for a while after this function returns.
+			// nolint
+			ctxDownload, _ := context.WithTimeout(ctx, downloadTimeout)
+			go downloadBinlogFilesForInstance(ctxDownload, instance, s.server.profile.DataDir, s.server.mysqlutil)
```

Comment:
This spawns additional go routines if the previous download isn't completed.

Reply:
The purpose is to spawn many concurrent goroutines to download binlog files for different instances.
Do you mean that we should set a max number of ongoing goroutines here? Or we only allow one downloading goroutine?